File: backend/services/omniaDialogStep/dialogStepEngine.js
/**
 * Motore runtime dialogo KB: filtra tabella, UC acquisition/inform/correction/complete.
 */

'use strict';

const {
  slugifyColumnId,
  normalizeCellValue,
  headerIndex,
  distinctColumnValues,
  filterRowsByBinding,
  listAskableColumns,
  fillInvalidationTemplate,
} = require('./kbDialogBindings');
const {
  resolveCompleteSay,
  resolveAcquisitionSay,
  findCorrectionIncompatibilities,
  resolveCorrectionMessages,
  isCorrectionUpdate,
} = require('./kbDialogSayResolver');
const {
  isEmptySelectorValue,
  computeInformKey,
  parseInformResponse,
  cloneInformState,
  invalidateInformForColumns,
  emptyInformState,
} = require('./kbDialogSelectorSemantics');
const {
  resolveInformSay,
  columnWantsInform,
  hasInformTemplate,
  DEFAULT_REJECT_SAY,
} = require('./kbDialogInformResolver');

function normalizeUpdates(updates, headers) {
  const out = {};
  if (!updates || typeof updates !== 'object') return out;
  for (const [key, raw] of Object.entries(updates)) {
    if (key === '__inform_response') continue;
    const val = normalizeCellValue(raw);
    if (!val) continue;
    const idx = headerIndex(headers, key);
    if (idx < 0) continue;
    out[slugifyColumnId(headers[idx])] = val;
  }
  return out;
}

function bindingKeysCanonical(binding, headers) {
  const out = {};
  for (const [key, val] of Object.entries(binding || {})) {
    const idx = headerIndex(headers, key);
    if (idx < 0) continue;
    const v = normalizeCellValue(val);
    if (!v) continue;
    out[slugifyColumnId(headers[idx])] = v;
  }
  return out;
}

function listSelectorColumns(selectorSpec) {
  if (!selectorSpec || !Array.isArray(selectorSpec.columns)) return [];
  return selectorSpec.columns
    .filter((c) => c.role === 'selector')
    .sort((a, b) => a.sortOrder - b.sortOrder || String(a.headerLabel).localeCompare(b.headerLabel, 'it'));
}

/** Autofill solo valori vuoti (nessun disclosure). */
function applyEmptyAutofills(binding, headers, rows, selectorSpec) {
  const next = { ...binding };
  for (const col of listSelectorColumns(selectorSpec)) {
    const colId = slugifyColumnId(col.headerLabel);
    if (normalizeCellValue(next[colId])) continue;
    const idx = headerIndex(headers, col.headerLabel);
    if (idx < 0) continue;
    const filtered = filterRowsByBinding(rows, headers, next);
    const distinct = distinctColumnValues(filtered, idx);
    if (distinct.length === 1 && isEmptySelectorValue(colId, distinct[0])) {
      next[colId] = distinct[0];
    }
  }
  return next;
}

/** Autofill tecnico valori significativi univoci (senza inform). */
function applyTechnicalAutofills(binding, headers, rows, selectorSpec, skipColIds = new Set()) {
  const next = { ...binding };
  for (const col of listSelectorColumns(selectorSpec)) {
    const colId = slugifyColumnId(col.headerLabel);
    if (skipColIds.has(colId) || normalizeCellValue(next[colId])) continue;
    const idx = headerIndex(headers, col.headerLabel);
    if (idx < 0) continue;
    const filtered = filterRowsByBinding(rows, headers, next);
    const distinct = distinctColumnValues(filtered, idx);
    if (distinct.length === 1 && !isEmptySelectorValue(colId, distinct[0])) {
      next[colId] = distinct[0];
    }
  }
  return next;
}

function pickInvalidationTemplate(selectorSpec) {
  const templates = selectorSpec?.invalidationTemplates ?? [];
  const approved = templates.find((t) => t.approved && t.template?.trim());
  if (approved) return approved;
  return templates.find((t) => t.template?.trim()) ?? null;
}

function rowToObject(row, headers) {
  const out = {};
  for (let i = 0; i < headers.length; i += 1) {
    const h = String(headers[i] ?? '').trim();
    if (!h) continue;
    out[slugifyColumnId(h)] = normalizeCellValue(row[i] ?? '') || '-';
    out[h] = out[slugifyColumnId(h)];
  }
  return out;
}

function buildBindingPrefixForCol(binding, colId, askable) {
  const out = {};
  for (const col of askable) {
    const id = slugifyColumnId(col.headerLabel);
    if (id === colId) break;
    const v = normalizeCellValue(binding[id] ?? '');
    if (v) out[id] = v;
  }
  return out;
}

function findNextInformAction(params) {
  const { binding, headers, rows, selectorSpec, dialogIndex, informState, askable } = params;
  const filtered = filterRowsByBinding(rows, headers, binding);
  const matched = filtered.length === 1 ? filtered[0] : filtered[0] ?? null;

  for (const col of listSelectorColumns(selectorSpec)) {
    const colId = slugifyColumnId(col.headerLabel);
    const idx = headerIndex(headers, col.headerLabel);
    if (idx < 0) continue;

    const currentVal = normalizeCellValue(binding[colId] ?? '');
    const prevDisclosed = informState.lastDisclosed[colId];
    const prevValue = prevDisclosed?.value;

    if (!currentVal) {
      if (prevValue && !isEmptySelectorValue(colId, prevValue)) {
        if (!columnWantsInform(col) || !hasInformTemplate(dialogIndex, colId)) continue;
        const informKey = computeInformKey(buildBindingPrefixForCol(binding, colId, askable), colId, prevValue);
        if (informState.acknowledged.includes(informKey)) continue;
        return {
          col,
          colId,
          value: '',
          prevValue,
          transitionKind: 'de_disclosure',
          matchedRow: matched,
          informKey,
        };
      }
      continue;
    }

    if (isEmptySelectorValue(colId, currentVal)) continue;
    if (!columnWantsInform(col) || !hasInformTemplate(dialogIndex, colId)) continue;

    const prefix = buildBindingPrefixForCol(binding, colId, askable);
    const informKey = computeInformKey(prefix, colId, currentVal);
    if (informState.acknowledged.includes(informKey)) continue;

    let transitionKind = 'disclosure';
    if (prevValue && !isEmptySelectorValue(colId, prevValue)) {
      if (prevValue.toLowerCase() === currentVal.toLowerCase()) continue;
      transitionKind = 'transition';
    } else if (prevValue && isEmptySelectorValue(colId, prevValue)) {
      transitionKind = 'disclosure';
    }

    const rowFiltered = filterRowsByBinding(rows, headers, binding);
    const rowDistinct = distinctColumnValues(rowFiltered, idx);
    if (rowDistinct.length !== 1 && !prevValue) continue;

    return {
      col,
      colId,
      value: currentVal,
      prevValue,
      transitionKind,
      matchedRow: matched,
      informKey,
    };
  }
  return null;
}

function baseResult(fields) {
  return fields;
}

/**
 * @param {object} params
 * @param {{ headers: string[], rows: string[][] }} params.grid
 * @param {object} params.selectorSpec
 * @param {Record<string, string>} params.binding
 * @param {Record<string, string>} [params.updates]
 * @param {object|null} [params.dialogIndex]
 * @param {object} [params.informState]
 */
function executeDialogStep({
  grid,
  selectorSpec,
  binding,
  updates,
  dialogIndex,
  informState: rawInformState,
}) {
  const headers = grid.headers;
  const rows = grid.rows;
  const askable = listAskableColumns(selectorSpec);
  let informState = cloneInformState(rawInformState ?? emptyInformState());

  const informResponse = parseInformResponse(updates);
  if (informState.informPending && informResponse) {
    const pending = informState.informPending;
    if (informResponse === 'reject') {
      informState.informPending = null;
      const cleared = { ...bindingKeysCanonical(binding, headers) };
      delete cleared[pending.colId];
      for (const col of askable) {
        const id = slugifyColumnId(col.headerLabel);
        if (col.sortOrder > (askable.find((c) => slugifyColumnId(c.headerLabel) === pending.colId)?.sortOrder ?? 999)) {
          delete cleared[id];
        }
      }
      informState = invalidateInformForColumns(informState, [pending.colId]);
      return baseResult({
        status: 'rejected',
        say: DEFAULT_REJECT_SAY,
        useCaseKind: 'inform',
        useCaseId: pending.useCaseId ?? null,
        binding: cleared,
        informState,
        conversationAction: 'terminate',
        remainingRowCount: filterRowsByBinding(rows, headers, cleared).length,
      });
    }
    if (informResponse === 'accept') {
      informState.acknowledged = [...new Set([...informState.acknowledged, pending.informKey])];
      informState.lastDisclosed[pending.colId] = { value: pending.value, say: pending.say };
      informState.informPending = null;
      return executeDialogStep({
        grid,
        selectorSpec,
        binding: bindingKeysCanonical(binding, headers),
        updates: {},
        dialogIndex,
        informState,
      });
    }
  }

  const priorBinding = bindingKeysCanonical(binding, headers);
  const normalizedUpdates = normalizeUpdates(updates, headers);
  const updateKeys = Object.keys(normalizedUpdates);

  let merged = { ...priorBinding };
  const rowsBefore = filterRowsByBinding(rows, headers, merged);

  for (const colId of updateKeys) {
    merged[colId] = normalizedUpdates[colId];
  }

  let filtered = filterRowsByBinding(rows, headers, merged);

  const correctionTrigger = isCorrectionUpdate(priorBinding, normalizedUpdates);
  if (correctionTrigger && dialogIndex) {
    const selectorColumns = listSelectorColumns(selectorSpec);
    const incompatibles = findCorrectionIncompatibilities(
      priorBinding,
      merged,
      headers,
      rows,
      selectorColumns,
      correctionTrigger
    );
    if (incompatibles.length > 0) {
      const { messages, binding: clearedBinding } = resolveCorrectionMessages(
        dialogIndex,
        correctionTrigger,
        incompatibles,
        merged,
        filtered,
        headers,
        rows,
        dialogIndex.valueLabels ?? {}
      );
      const clearedIds = incompatibles.map((i) => i.columnId).filter(Boolean);
      informState = invalidateInformForColumns(informState, clearedIds);
      merged = applyEmptyAutofills(clearedBinding, headers, rows, selectorSpec);
      merged = applyTechnicalAutofills(merged, headers, rows, selectorSpec);
      filtered = filterRowsByBinding(rows, headers, merged);
      return baseResult({
        status: 'correction',
        say: messages.filter(Boolean).join(' '),
        binding: merged,
        useCaseKind: 'correction',
        informState,
        remainingRowCount: filtered.length,
        corrections: incompatibles,
      });
    }
  }

  if (filtered.length === 0 && rowsBefore.length > 0 && updateKeys.length > 0) {
    const rejectedColId = updateKeys[updateKeys.length - 1];
    const rejectedVal = merged[rejectedColId];
    const partial = { ...merged };
    delete partial[rejectedColId];
    const partialRows = filterRowsByBinding(rows, headers, partial);
    const col = askable.find((c) => slugifyColumnId(c.headerLabel) === rejectedColId);
    const colIdx = headerIndex(headers, rejectedColId);
    const alternatives = colIdx >= 0 ? distinctColumnValues(partialRows, colIdx) : [];
    const alt =
      alternatives.find((a) => a.toLowerCase() !== String(rejectedVal).toLowerCase()) ??
      alternatives[0] ??
      '';

    const tpl = pickInvalidationTemplate(selectorSpec);
    const colLabel = col?.promptTemplate ?? rejectedColId;
    const say = tpl
      ? fillInvalidationTemplate(tpl.template, {
          colonna: colLabel,
          valore_rifiutato: rejectedVal,
          alternativa_suggerita: alt,
          alternativa: alt,
          [rejectedColId]: rejectedVal,
          specialita: partial.specialita ?? merged.specialita ?? '',
          tipo_visita: partial.tipo_visita ?? merged.tipo_visita ?? '',
        })
      : `La combinazione scelta non è disponibile. ${alt ? `Può andare bene ${alt}.` : ''}`;

    return baseResult({
      status: 'invalid',
      say: say.trim(),
      binding: partial,
      informState,
      rejected: { columnId: rejectedColId, value: rejectedVal, alternative: alt },
      remainingRowCount: partialRows.length,
      allowedValues: alternatives,
    });
  }

  merged = applyEmptyAutofills(merged, headers, rows, selectorSpec);
  filtered = filterRowsByBinding(rows, headers, merged);

  const pending = askable.filter((col) => {
    const colId = slugifyColumnId(col.headerLabel);
    return !normalizeCellValue(merged[colId]);
  });

  const nextCol = pending[0];
  if (nextCol) {
    const nextColId = slugifyColumnId(nextCol.headerLabel);
    const colIdx = headerIndex(headers, nextCol.headerLabel);
    const allowedValues = colIdx >= 0 ? distinctColumnValues(filtered, colIdx) : [];

    if (allowedValues.length === 1) {
      const single = allowedValues[0];
      if (isEmptySelectorValue(nextColId, single)) {
        merged[nextColId] = single;
        filtered = filterRowsByBinding(rows, headers, merged);
      } else if (columnWantsInform(nextCol) && hasInformTemplate(dialogIndex, nextColId)) {
        merged[nextColId] = single;
        filtered = filterRowsByBinding(rows, headers, merged);
      } else {
        merged[nextColId] = single;
        return executeDialogStep({
          grid,
          selectorSpec,
          binding: merged,
          updates: {},
          dialogIndex,
          informState,
        });
      }
    }
  }

  merged = applyTechnicalAutofills(merged, headers, rows, selectorSpec);
  filtered = filterRowsByBinding(rows, headers, merged);

  const informAction = findNextInformAction({
    binding: merged,
    headers,
    rows,
    selectorSpec,
    dialogIndex,
    informState,
    askable,
  });

  if (informAction) {
    const resolved = resolveInformSay({
      dialogIndex,
      col: informAction.col,
      colId: informAction.colId,
      binding: merged,
      value: informAction.value,
      prevValue: informAction.prevValue,
      matchedRow: informAction.matchedRow,
      headers,
      transitionKind: informAction.transitionKind,
    });

    if (resolved.requiresAcceptance) {
      informState.informPending = {
        colId: informAction.colId,
        value: informAction.value,
        informKey: informAction.informKey,
        useCaseId: resolved.useCaseId,
        say: resolved.say,
      };
      return baseResult({
        status: 'inform_pending',
        say: resolved.say,
        useCaseId: resolved.useCaseId,
        useCaseKind: 'inform',
        binding: merged,
        informState,
        requiresAcceptance: true,
        informColumnId: informAction.colId,
        remainingRowCount: filtered.length,
      });
    }

    informState.acknowledged = [...new Set([...informState.acknowledged, informAction.informKey])];
    if (informAction.value) {
      informState.lastDisclosed[informAction.colId] = {
        value: informAction.value,
        say: resolved.say,
      };
    } else {
      delete informState.lastDisclosed[informAction.colId];
    }

    return baseResult({
      status: 'inform',
      say: resolved.say,
      useCaseId: resolved.useCaseId,
      useCaseKind: 'inform',
      binding: merged,
      informState,
      requiresAcceptance: false,
      informColumnId: informAction.colId,
      remainingRowCount: filtered.length,
    });
  }

  const pendingAfter = askable.filter((col) => {
    const colId = slugifyColumnId(col.headerLabel);
    return !normalizeCellValue(merged[colId]);
  });

  if (pendingAfter.length === 0 || filtered.length <= 1) {
    const matched = filtered[0] ?? null;
    if (matched && dialogIndex) {
      const resolved = resolveCompleteSay(dialogIndex, merged, headers, matched);
      return baseResult({
        status: filtered.length > 0 ? 'complete' : 'error',
        say: resolved.say,
        sayCore: resolved.sayCore,
        useCaseId: resolved.useCaseId,
        useCaseKind: resolved.useCaseKind,
        binding: merged,
        informState,
        remainingRowCount: filtered.length,
        matchedRow: rowToObject(matched, headers),
        matchedRows: filtered.map((r) => rowToObject(r, headers)),
      });
    }
    const say = matched
      ? 'Perfetto, ho trovato la combinazione disponibile.'
      : 'Non ho trovato combinazioni disponibili con le scelte fatte.';
    return baseResult({
      status: filtered.length > 0 ? 'complete' : 'error',
      say,
      binding: merged,
      informState,
      remainingRowCount: filtered.length,
      matchedRow: matched ? rowToObject(matched, headers) : null,
      matchedRows: filtered.map((r) => rowToObject(r, headers)),
    });
  }

  const askCol = pendingAfter[0];
  const askColIdx = headerIndex(headers, askCol.headerLabel);
  const allowedValues = askColIdx >= 0 ? distinctColumnValues(filtered, askColIdx) : [];
  const askColId = slugifyColumnId(askCol.headerLabel);

  const acquired = dialogIndex
    ? resolveAcquisitionSay(dialogIndex, askColId, merged, askCol, allowedValues)
    : null;
  const say = acquired?.say ?? `${askCol.promptTemplate ?? askCol.headerLabel}?`;

  return baseResult({
    status: 'ask',
    say,
    useCaseId: acquired?.useCaseId ?? null,
    useCaseKind: acquired?.useCaseKind ?? 'acquisition',
    binding: merged,
    informState,
    nextColumnId: askColId,
    nextHeaderLabel: askCol.headerLabel,
    promptType: askCol.promptType,
    askPolicy: askCol.askPolicy,
    allowedValues,
    remainingRowCount: filtered.length,
  });
}

module.exports = {
  executeDialogStep,
  normalizeUpdates,
  bindingKeysCanonical,
};

File: VBNET/Engine/KbDialogStep/DialogStepEngine.vb
Option Strict On
Option Explicit On

Imports System.Collections.Generic
Imports System.Globalization
Imports System.Linq
Imports Newtonsoft.Json.Linq

Namespace KbDialogStep
    ''' <summary>Motore runtime dialogo KB: filtra tabella, UC acquisition/correction/complete.</summary>
    Public NotInheritable Class DialogStepEngine
        Private Shared ReadOnly ItalianComparer As StringComparer = StringComparer.Create(New CultureInfo("it-IT"), ignoreCase:=False)

        Public Shared Function NormalizeUpdates(updates As Dictionary(Of String, String), headers As IList(Of String)) As Dictionary(Of String, String)
            Dim out As New Dictionary(Of String, String)(StringComparer.OrdinalIgnoreCase)
            If updates Is Nothing OrElse headers Is Nothing Then Return out
            For Each kvp In updates
                Dim val = KbDialogBindings.NormalizeCellValue(kvp.Value)
                If val.Length = 0 Then Continue For
                Dim idx = KbDialogBindings.HeaderIndex(headers, kvp.Key)
                If idx < 0 Then Continue For
                out(KbDialogBindings.SlugifyColumnId(headers(idx))) = val
            Next
            Return out
        End Function

        Public Shared Function BindingKeysCanonical(binding As Dictionary(Of String, String), headers As IList(Of String)) As Dictionary(Of String, String)
            Dim out As New Dictionary(Of String, String)(StringComparer.OrdinalIgnoreCase)
            If binding Is Nothing OrElse headers Is Nothing Then Return out
            For Each kvp In binding
                Dim idx = KbDialogBindings.HeaderIndex(headers, kvp.Key)
                If idx < 0 Then Continue For
                Dim v = KbDialogBindings.NormalizeCellValue(kvp.Value)
                If v.Length = 0 Then Continue For
                out(KbDialogBindings.SlugifyColumnId(headers(idx))) = v
            Next
            Return out
        End Function

        Private Shared Function ListSelectorColumns(selectorSpec As KbSelectorSpec) As List(Of SelectorColumnSpec)
            If selectorSpec Is Nothing OrElse selectorSpec.Columns Is Nothing Then Return New List(Of SelectorColumnSpec)
            Return selectorSpec.Columns.
                Where(Function(c) c IsNot Nothing AndAlso String.Equals(c.Role, "selector", StringComparison.OrdinalIgnoreCase)).
                OrderBy(Function(c) c.SortOrder).
                ThenBy(Function(c) If(c.HeaderLabel, ""), ItalianComparer).
                ToList()
        End Function

        Private Shared Function ApplyAutoFills(binding As Dictionary(Of String, String), headers As IList(Of String), rows As IEnumerable(Of IList(Of String)), selectorSpec As KbSelectorSpec) As Dictionary(Of String, String)
            Dim nextBinding As New Dictionary(Of String, String)(StringComparer.OrdinalIgnoreCase)
            If binding IsNot Nothing Then
                For Each kvp In binding
                    nextBinding(kvp.Key) = kvp.Value
                Next
            End If
            For Each col In ListSelectorColumns(selectorSpec)
                Dim colId = KbDialogBindings.SlugifyColumnId(col.HeaderLabel)
                If KbDialogBindings.NormalizeCellValue(If(nextBinding.ContainsKey(colId), nextBinding(colId), "")).Length > 0 Then Continue For
                Dim idx = KbDialogBindings.HeaderIndex(headers, col.HeaderLabel)
                If idx < 0 Then Continue For
                Dim filtered = KbDialogBindings.FilterRowsByBinding(rows, headers, nextBinding)
                Dim distinct = KbDialogBindings.DistinctColumnValues(filtered, idx)
                If distinct.Count = 1 Then nextBinding(colId) = distinct(0)
            Next
            Return nextBinding
        End Function

        Private Shared Function PickInvalidationTemplate(selectorSpec As KbSelectorSpec) As InvalidationTemplateSpec
            Dim templates = If(selectorSpec IsNot Nothing AndAlso selectorSpec.InvalidationTemplates IsNot Nothing,
                               selectorSpec.InvalidationTemplates,
                               New List(Of InvalidationTemplateSpec))
            Dim approved = templates.FirstOrDefault(Function(t) t IsNot Nothing AndAlso t.Approved AndAlso If(t.Template, "").Trim().Length > 0)
            If approved IsNot Nothing Then Return approved
            Return templates.FirstOrDefault(Function(t) t IsNot Nothing AndAlso If(t.Template, "").Trim().Length > 0)
        End Function

        Private Shared Function RowToObject(row As IList(Of String), headers As IList(Of String)) As Dictionary(Of String, String)
            Dim out As New Dictionary(Of String, String)(StringComparer.OrdinalIgnoreCase)
            For i = 0 To headers.Count - 1
                Dim h = If(headers(i), "").Trim()
                If h.Length = 0 Then Continue For
                Dim slug = KbDialogBindings.SlugifyColumnId(h)
                Dim val = KbDialogBindings.NormalizeCellValue(If(i < row.Count, row(i), ""))
                If val.Length = 0 Then val = "-"
                out(slug) = val
                out(h) = val
            Next
            Return out
        End Function

        Private Shared Function CopyBinding(binding As Dictionary(Of String, String)) As Dictionary(Of String, String)
            Dim copy As New Dictionary(Of String, String)(StringComparer.OrdinalIgnoreCase)
            If binding Is Nothing Then Return copy
            For Each kvp In binding
                copy(kvp.Key) = kvp.Value
            Next
            Return copy
        End Function

        Public Shared Function ExecuteDialogStep(grid As KbDialogGrid, selectorSpec As KbSelectorSpec, binding As Dictionary(Of String, String), Optional updates As Dictionary(Of String, String) = Nothing, Optional dialogIndex As JObject = Nothing, Optional informState As DialogInformState = Nothing) As DialogStepResult
            If informState Is Nothing Then informState = KbDialogSelectorSemantics.EmptyInformState()
            Dim headers = grid.Headers
            Dim rows = grid.Rows
            Dim askable = KbDialogBindings.ListAskableColumns(selectorSpec)

            Dim priorBinding = BindingKeysCanonical(binding, headers)
            Dim normalizedUpdates = NormalizeUpdates(updates, headers)
            Dim updateKeys = normalizedUpdates.Keys.ToList()

            Dim merged = CopyBinding(priorBinding)
            Dim rowsBefore = KbDialogBindings.FilterRowsByBinding(rows, headers, merged)

            For Each colId In updateKeys
                merged(colId) = normalizedUpdates(colId)
            Next

            Dim filtered = KbDialogBindings.FilterRowsByBinding(rows, headers, merged)

            Dim correctionTrigger = KbDialogSayResolver.IsCorrectionUpdate(priorBinding, normalizedUpdates)
            If correctionTrigger IsNot Nothing AndAlso dialogIndex IsNot Nothing Then
                Dim incompatibles = KbDialogSayResolver.FindCorrectionIncompatibilities(
                    priorBinding, merged, headers, rows, ListSelectorColumns(selectorSpec), correctionTrigger)
                If incompatibles.Count > 0 Then
                    Dim resolved = KbDialogSayResolver.ResolveCorrectionMessages(
                        dialogIndex, correctionTrigger, incompatibles, merged, headers, rows, dialogIndex("valueLabels"))
                    merged = ApplyAutoFills(resolved.Binding, headers, rows, selectorSpec)
                    filtered = KbDialogBindings.FilterRowsByBinding(rows, headers, merged)
                    Dim correctionSay = String.Join(" ", resolved.Messages.Where(Function(m) Not String.IsNullOrWhiteSpace(m)))
                    Dim correctionUseCaseId As String = Nothing
                    Dim correctionArr = dialogIndex("correction")
                    If correctionArr IsNot Nothing AndAlso correctionArr.Type = JTokenType.Array Then
                        For Each item In CType(correctionArr, JArray)
                            If String.Equals(item.Value(Of String)("triggerColumnId"), correctionTrigger, StringComparison.OrdinalIgnoreCase) Then
                                correctionUseCaseId = item.Value(Of String)("useCaseId")
                                Exit For
                            End If
                        Next
                    End If
                    Return New DialogStepResult With {
                        .Status = "correction",
                        .Say = correctionSay.Trim(),
                        .UseCaseId = correctionUseCaseId,
                        .UseCaseKind = "correction",
                        .Binding = merged,
                        .RemainingRowCount = filtered.Count
                    }
                End If
            End If

            If filtered.Count = 0 AndAlso rowsBefore.Count > 0 AndAlso updateKeys.Count > 0 Then
                Dim rejectedColId = updateKeys(updateKeys.Count - 1)
                Dim rejectedVal = merged(rejectedColId)
                Dim partialBinding = CopyBinding(merged)
                partialBinding.Remove(rejectedColId)
                Dim partialRows = KbDialogBindings.FilterRowsByBinding(rows, headers, partialBinding)
                Dim col = askable.FirstOrDefault(Function(c) KbDialogBindings.SlugifyColumnId(c.HeaderLabel) = rejectedColId)
                Dim colIdx = KbDialogBindings.HeaderIndex(headers, rejectedColId)
                Dim alternatives = If(colIdx >= 0, KbDialogBindings.DistinctColumnValues(partialRows, colIdx), New List(Of String))
                Dim alt = alternatives.FirstOrDefault(Function(a) Not String.Equals(a, rejectedVal, StringComparison.OrdinalIgnoreCase))
                If alt Is Nothing AndAlso alternatives.Count > 0 Then alt = alternatives(0)
                If alt Is Nothing Then alt = ""

                Dim tpl = PickInvalidationTemplate(selectorSpec)
                Dim colLabel = If(col IsNot Nothing AndAlso Not String.IsNullOrEmpty(col.PromptTemplate), col.PromptTemplate, rejectedColId)
                Dim say As String
                If tpl IsNot Nothing AndAlso Not String.IsNullOrWhiteSpace(tpl.Template) Then
                    Dim vars As New Dictionary(Of String, String) From {
                        {"colonna", colLabel},
                        {"valore_rifiutato", rejectedVal},
                        {"alternativa_suggerita", alt},
                        {"alternativa", alt}
                    }
                    vars(rejectedColId) = rejectedVal
                    Dim specialita As String = ""
                    If partialBinding.ContainsKey("specialita") Then
                        specialita = partialBinding("specialita")
                    ElseIf merged.ContainsKey("specialita") Then
                        specialita = merged("specialita")
                    End If
                    Dim tipoVisita As String = ""
                    If partialBinding.ContainsKey("tipo_visita") Then
                        tipoVisita = partialBinding("tipo_visita")
                    ElseIf merged.ContainsKey("tipo_visita") Then
                        tipoVisita = merged("tipo_visita")
                    End If
                    vars("specialita") = specialita
                    vars("tipo_visita") = tipoVisita
                    say = KbDialogBindings.FillInvalidationTemplate(tpl.Template, vars)
                Else
                    say = "La combinazione scelta non è disponibile." & If(alt.Length > 0, " Può andare bene " & alt & ".", "")
                End If

                Return New DialogStepResult With {
                    .Status = "invalid",
                    .Say = say.Trim(),
                    .Binding = partialBinding,
                    .Rejected = New DialogStepRejectedInfo With {.ColumnId = rejectedColId, .Value = rejectedVal, .Alternative = alt},
                    .RemainingRowCount = partialRows.Count,
                    .AllowedValues = alternatives
                }
            End If

            merged = ApplyAutoFills(merged, headers, rows, selectorSpec)
            filtered = KbDialogBindings.FilterRowsByBinding(rows, headers, merged)

            Dim pending = askable.Where(Function(c)
                                            Dim colId = KbDialogBindings.SlugifyColumnId(c.HeaderLabel)
                                            Return KbDialogBindings.NormalizeCellValue(If(merged.ContainsKey(colId), merged(colId), "")).Length = 0
                                        End Function).ToList()

            If pending.Count = 0 OrElse filtered.Count <= 1 Then
                Dim matched = If(filtered.Count > 0, filtered(0), Nothing)
                If matched IsNot Nothing AndAlso dialogIndex IsNot Nothing Then
                    Dim resolved = KbDialogSayResolver.ResolveCompleteSay(dialogIndex, merged, headers, matched)
                    Return New DialogStepResult With {
                        .Status = If(filtered.Count > 0, "complete", "error"),
                        .Say = resolved.Say,
                        .SayCore = resolved.SayCore,
                        .UseCaseId = resolved.UseCaseId,
                        .UseCaseKind = resolved.UseCaseKind,
                        .Binding = merged,
                        .RemainingRowCount = filtered.Count,
                        .MatchedRow = RowToObject(matched, headers),
                        .MatchedRows = filtered.Select(Function(r) RowToObject(r, headers)).ToList()
                    }
                End If
                Dim sayComplete = If(matched IsNot Nothing,
                    "Perfetto, ho trovato la combinazione disponibile.",
                    "Non ho trovato combinazioni disponibili con le scelte fatte.")
                Return New DialogStepResult With {
                    .Status = If(filtered.Count > 0, "complete", "error"),
                    .Say = sayComplete,
                    .Binding = merged,
                    .RemainingRowCount = filtered.Count,
                    .MatchedRow = If(matched IsNot Nothing, RowToObject(matched, headers), Nothing),
                    .MatchedRows = filtered.Select(Function(r) RowToObject(r, headers)).ToList()
                }
            End If

            Dim nextCol = pending(0)
            Dim nextColIdx = KbDialogBindings.HeaderIndex(headers, nextCol.HeaderLabel)
            Dim allowedValues = If(nextColIdx >= 0, KbDialogBindings.DistinctColumnValues(filtered, nextColIdx), New List(Of String))
            Dim nextColId = KbDialogBindings.SlugifyColumnId(nextCol.HeaderLabel)

            If allowedValues.Count = 1 Then
                merged(nextColId) = allowedValues(0)
                Return ExecuteDialogStep(grid, selectorSpec, merged, New Dictionary(Of String, String), dialogIndex, informState)
            End If

            Dim acquired As KbDialogSayResolver.AcquisitionResolveResult = Nothing
            If dialogIndex IsNot Nothing Then
                acquired = KbDialogSayResolver.ResolveAcquisitionSay(dialogIndex, nextColId, merged, nextCol, allowedValues)
            End If
            Dim sayAsk = If(acquired IsNot Nothing AndAlso Not String.IsNullOrWhiteSpace(acquired.Say),
                            acquired.Say,
                            If(Not String.IsNullOrEmpty(nextCol.PromptTemplate), nextCol.PromptTemplate, nextCol.HeaderLabel) & "?")

            Return New DialogStepResult With {
                .Status = "ask",
                .Say = sayAsk,
                .UseCaseId = If(acquired?.UseCaseId, Nothing),
                .UseCaseKind = If(acquired?.UseCaseKind, "acquisition"),
                .Binding = merged,
                .InformState = informState,
                .NextColumnId = nextColId,
                .NextHeaderLabel = nextCol.HeaderLabel,
                .PromptType = nextCol.PromptType,
                .AskPolicy = nextCol.AskPolicy,
                .AllowedValues = allowedValues,
                .RemainingRowCount = filtered.Count
            }
        End Function
    End Class
End Namespace
